python/ql/src/experimental/Security/CWE-348/ClientSuppliedIpUsedInSecurityCheck.ql

```diff
@@ -0,0 +1,57 @@
+/**
+ * @name IP address spoofing
+ * @description A remote endpoint identifier is read from an HTTP header. Attackers can modify the value
+ *              of the identifier to forge the client ip.
+ * @kind path-problem
+ * @problem.severity error
+ * @precision high
+ * @id py/ip-address-spoofing
+ * @tags security
+ *       external/cwe/cwe-348
+ */
+
+import python
+import semmle.python.dataflow.new.DataFlow
+import semmle.python.dataflow.new.TaintTracking
+import semmle.python.ApiGraphs
+import ClientSuppliedIpUsedInSecurityCheckLib
+import DataFlow::PathGraph
+
+/**
+ * Taint-tracking configuration tracing flow from obtaining a client ip from an HTTP header to a sensitive use.
+ */
+class ClientSuppliedIpUsedInSecurityCheckConfig extends TaintTracking::Configuration {
+  ClientSuppliedIpUsedInSecurityCheckConfig() { this = "ClientSuppliedIpUsedInSecurityCheckConfig" }
+
+  override predicate isSource(DataFlow::Node source) {
+    source instanceof ClientSuppliedIpUsedInSecurityCheck
+  }
+
+  override predicate isSink(DataFlow::Node sink) {
+    sink instanceof ClientSuppliedIpUsedInSecurityCheckSink
+  }
+
+  override predicate isAdditionalTaintStep(DataFlow::Node pred, DataFlow::Node succ) {
+    exists(DataFlow::CallCfgNode ccn |
+      ccn = API::moduleImport("netaddr").getMember("IPAddress").getACall() and
+      ccn.getArg(0) = pred and
+      ccn = succ
+    )
+  }
+
+  override predicate isSanitizer(DataFlow::Node node) {
+    exists(Subscript ss |
+      not ss.getIndex().(IntegerLiteral).getText() = "0" and
+      ss.getObject().(Call).getFunc().(Attribute).getName() = "split" and
+      ss.getObject().(Call).getArg(0).(StrConst).getText() = "," and
+      ss.getObject().(Call).getFunc().(Attribute).getObject() = node.asExpr()
+    )
+  }
+}
```

Comment:
Should the sanitizer (`node`) not be the result of the subscripting rather than object? Consider the code
```python
  client_ip = request.headers.get('x-forwarded-for')
  dummy = client_ip.split(",")[2]
  if not client_ip == '127.0.0.1':
         raise Exception('ip illegal')
```
This would be deemed sanitized as the flow is blocked in the line assigning `dummy`. Yet, no change was made to `client_ip` prior to testing it. (If the sanitizer is the subscript, the flow is blocked into `dummy`, but there is still use-use-flow to the variable on the next line.)

Reply:
Is there such a way of writing in real open source projects?

Reply:
FYI: https://github.com/github/codeql/blob/main/java/ql/src/experimental/Security/CWE/CWE-348/ClientSuppliedIpUsedInSecurityCheck.ql#L37-L41

Reply:
X-Forwarded-For is a superposition process when used, format: client, proxy1, proxy2. X-Real-IP is a process of coverage when in use.

Reply:
It is possible that we have found a possible improvement to the java query :-) While this particular pattern looks unlikely, I think it could easily happen as a typo, at which point an alert would be great. And other patterns could be constructed as well, so we should always strive to point to the correct node. In any case, it should not be hard to fix.

Reply:
OK. Can you elaborate on this disinfectant? Let me optimize the configuration.

Reply:
```ql
override predicate isSanitizer(DataFlow::Node node) {
    exists(Subscript ss |
      ss.getObject().(Call).getFunc().(Attribute).getName() = "split" and
      ss.getObject().(Call).getArg(0).(StrConst).getText() = "," and
      ss = node.asExpr()
    )
  }
```
Do you want this result?

Reply:
Yes exactly. Except I think you still need to exclude index 0.